Exclude hibernate-ehcache

jamesfredley opened the pull request:
resolves `hibernate-core` and `hibernate-core-jakarta` being on the classpath

https://github.com/grails/gorm-hibernate5/issues/951

jamesfredley committed:
exclude hibernate-core dependency

jamesfredley committed:
add jboss-transaction-api dependency

jamesfredley commented on grails-plugin/build.gradle:
```diff
@@ -25,7 +25,14 @@ dependencies {
+        exclude group:'org.hibernate', module:'hibernate-core'
+    }
+    // required for org.hibernate:hibernate-ehcache to work with org.hibernate:hibernate-core-jakarta
+    api "org.jboss.spec.javax.transaction:jboss-transaction-api_1.3_spec:$jbossTransactionApiVersion", {
```
This is suboptimal, but required to work around the following test failure.  It was coming from hibernate-core.

```
> Task :examples-grails-hibernate:test
 -- Executed test initializationError [functional.tests.BookControllerUnitSpec] with result: FAILURE

BookControllerUnitSpec > initializationError FAILED
    java.lang.ExceptionInInitializerError
        at org.hibernate.cache.ehcache.internal.EhcacheRegionFactory.<clinit>(EhcacheRegionFactory.java:46)
        at java.base/java.lang.Class.newInstance(Class.java:645)
        at org.hibernate.cache.internal.StrategyCreatorRegionFactoryImpl.create(StrategyCreatorRegionFactoryImpl.java:62)
        at org.hibernate.cache.internal.StrategyCreatorRegionFactoryImpl.create(StrategyCreatorRegionFactoryImpl.java:23)
        at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.resolveStrategy(StrategySelectorImpl.java:243)
        at org.hibernate.boot.registry.selector.internal.StrategySelectorImpl.resolveStrategy(StrategySelectorImpl.java:192)
        at org.hibernate.cache.internal.RegionFactoryInitiator.resolveRegionFactory(RegionFactoryInitiator.java:93)
        at org.hibernate.cache.internal.RegionFactoryInitiator.initiateService(RegionFactoryInitiator.java:47)
        at org.hibernate.cache.internal.RegionFactoryInitiator.initiateService(RegionFactoryInitiator.java:32)
        at org.hibernate.boot.registry.internal.StandardServiceRegistryImpl.initiateService(StandardServiceRegistryImpl.java:101)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:272)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.initializeService(AbstractServiceRegistryImpl.java:246)
        at org.hibernate.service.internal.AbstractServiceRegistryImpl.getService(AbstractServiceRegistryImpl.java:223)
        at org.hibernate.boot.internal.MetadataBuilderImpl$MetadataBuildingOptionsImpl.<init>(MetadataBuilderImpl.java:713)
        at org.hibernate.boot.internal.MetadataBuilderImpl.<init>(MetadataBuilderImpl.java:123)
        at org.hibernate.boot.MetadataSources.getMetadataBuilder(MetadataSources.java:158)
        at org.hibernate.cfg.Configuration.buildSessionFactory(Configuration.java:673)
        at org.grails.orm.hibernate.cfg.HibernateMappingContextConfiguration.buildSessionFactory(HibernateMappingContextConfiguration.java:287)
        at org.grails.orm.hibernate.connections.HibernateConnectionSourceFactory.create(HibernateConnectionSourceFactory.java:86)
        at org.grails.orm.hibernate.connections.AbstractHibernateConnectionSourceFactory.create(AbstractHibernateConnectionSourceFactory.java:39)
        at org.grails.orm.hibernate.connections.AbstractHibernateConnectionSourceFactory.create(AbstractHibernateConnectionSourceFactory.java:23)
        at org.grails.datastore.mapping.core.connections.AbstractConnectionSourceFactory.create(AbstractConnectionSourceFactory.java:64)
        at org.grails.datastore.mapping.core.connections.AbstractConnectionSourceFactory.create(AbstractConnectionSourceFactory.java:52)
        at org.grails.datastore.mapping.core.connections.ConnectionSourcesInitializer.create(ConnectionSourcesInitializer.groovy:24)
        at org.grails.orm.hibernate.HibernateDatastore.<init>(HibernateDatastore.java:222)
        at org.grails.orm.hibernate.HibernateDatastore.<init>(HibernateDatastore.java:276)
        at org.grails.orm.hibernate.HibernateDatastore.<init>(HibernateDatastore.java:285)
        at grails.test.hibernate.HibernateSpec.setupSpec(HibernateSpec.groovy:62)

        Caused by:
        java.lang.IllegalArgumentException: The given lookup does not have access to the implementation class constructor
            at org.jboss.logging.Logger.doGetMessageLogger(Logger.java:2641)
            at org.jboss.logging.Logger.getMessageLogger(Logger.java:2570)
            at org.jboss.logging.Logger.getMessageLogger(Logger.java:2552)
            at org.hibernate.cache.ehcache.internal.EhCacheMessageLogger.<clinit>(EhCacheMessageLogger.java:31)
            ... 28 more
```

jamesfredley committed:
remove doc change from Grails 3

matrei commented:
Couple of thoughts about this:
1. `org.hiberate:hibernate-ehcache` is, what I can see, not a dependency of either `grails-datastore-gorm-hibernate5` or `grails-plugin` and can be removed as such.
2. `org.hibernate:hibernate-ehcache` is a `testRuntimeOnly` dependency of `grails-plugin` as tests explicitly sets the `hibernate:cache:region.factory_class` to classes from `org.hibernate:hibernate-ehcache` in configuration.
3. `org.hibernate:hibernate-ehcache` is a `runtimeOnly` dependency of `examples-grails-hibernate` as it explicitly sets the `hibernate:cache:region.factory_class` to `org.hibernate.cache.ehcache.EhCacheRegionFactory` in `application.yml`.
4. Making `org.jboss.spec.javax.transaction:jboss-transaction-api_1.3_spec` a dependency of `grails-plugin` should not be needed.

I believe it is the responsibility of the consumer of `org.hibernate:hibernate-ehcache` to make sure it is on the classpath and that an incompatible version of `org.hibernate:hibernate-core` is excluded from it, if it causes conflicts.

@jamesfredley If you agree with these thoughts, would you like me to add them to this PR?

jamesfredley commented:
@matrei 

> org.hiberate:hibernate-ehcache is, what I can see, not a dependency of neither grails-datastore-gorm-hibernate5 or grails-plugin and can be removed as such.

This is true and it was removed around Sept of 2024, but then broke a number of tests in other Grails projects which depended upon it being provided by the hibernate5 plugin historically.  So it is a trade off of a breaking change for Grails apps using ehcache during Grails 7 upgrade vs removing a dependency that is not directly used by the plugin.  ehache does not exist for hibernate6, so when gorm-hibernate6 is ready, jcache will be part of the transition.  

> Making org.jboss.spec.javax.transaction:jboss-transaction-api_1.3_spec a dependency of grails-plugin should not be needed.

If we remove `org.hiberate:hibernate-ehcache` as a `grails-plugin` dependency, there are cases like https://github.com/grails/gorm-hibernate5/pull/955#discussion_r1941977033 where the end user will need to include it, like the following:

```
implementation "org.hibernate:hibernate-ehcache:$hibernateVersion", {
    exclude group:'org.hibernate', module:'hibernate-core'
}
// required for org.hibernate:hibernate-ehcache to work with org.hibernate:hibernate-core-jakarta
implementation "org.jboss.spec.javax.transaction:jboss-transaction-api_1.3_spec:$jbossTransactionApiVersion", {
    exclude group:"jakarta.enterprise", module:"jakarta.enterprise.cdi-api"
}

```

A quick start to the list of locations that will need to have `org.hibernate:hibernate-ehcache` added:

- https://github.com/search?q=org%3Agrails+org.hibernate.cache.ehcache+language%3AYAML&type=code&l=YAML

matrei commented:
> So it is a trade off of a breaking change for Grails apps using ehcache during Grails 7 upgrade vs removing a dependency that is not directly used by the plugin

I’d prefer a breaking change in a major version update rather than continuing with the wrong approach. Why should every Grails app using `org.grails.plugins:hibernate5` include this dependency, even if it’s not needed? This "kitchen-sink" approach adds unnecessary complexity and makes it harder for users to understand the libraries they depend on. I believe a cleaner, more intentional dependency structure benefits everyone in the long run.

> If we remove org.hiberate:hibernate-ehcache as a grails-plugin dependency, there are cases like https://github.com/grails/gorm-hibernate5/pull/955#discussion_r1941977033 where the end user will need to include it, like the following

I looked through the issues related to this PR but couldn’t find a project demonstrating the need for `org.jboss.spec.javax.transaction:jboss-transaction-api_1.3_spec`.

It seems that `BookControllerUnitSpec` works when adding `testRuntimeOnly 'org.hibernate:hibernate-ehcache'` to the `grails-plugin` project. What am I missing?

jamesfredley commented:
@matrei Ok, please proceed with the changes.

> It seems that BookControllerUnitSpec works when adding testRuntimeOnly 'org.hibernate:hibernate-ehcache' to the grails-plugin project. What am I missing?

That is excellent news, as I did not like adding that javax dependency.

I'll be glad to work on updating the other Grails projects after this breaking change.

matrei committed:
build: remove module dependency on `hibernate-ehcache`

Moved `hibernate-ehcache` to a runtime dependency where needed for tests, making it optional and reducing unnecessary module coupling.

As a result, applications using `hibernate-ehcache` will now need to explicitly add it as a dependency.

jdaugherty commented:
Can we update the upgrade guide for Grails 7 to include these instructions since it is a breaking change?

matrei commented:
> Can we update the upgrade guide for Grails 7 to include these instructions since it is a breaking change?

https://github.com/grails/grails-doc/pull/951

jamesfredley commented on examples/grails-hibernate/build.gradle:
```diff
@@ -26,6 +26,7 @@ dependencies {
     runtimeOnly 'org.grails:grails-plugin-services'
     runtimeOnly 'org.grails:grails-plugin-url-mappings'
     runtimeOnly 'org.grails.plugins:fields'
+    runtimeOnly "org.hibernate:hibernate-ehcache:$hibernateVersion"
```
Add the following to document the right approach and avoid conflicts from the javax version?

```
, {
    exclude group:'org.hibernate', module:'hibernate-core'
}
```

jamesfredley commented on grails-datastore-gorm-hibernate/build.gradle:
```diff
@@ -42,20 +29,17 @@ dependencies {
     testImplementation "org.apache.tomcat:tomcat-jdbc"
-    testRuntimeOnly "org.springframework:spring-aop"
 
+    testRuntimeOnly "org.hibernate:hibernate-ehcache:$hibernateVersion"
```
Add the following to document the right approach and avoid conflicts from the javax version?

```
, {
    exclude group:'org.hibernate', module:'hibernate-core'
}
```

jamesfredley commented on grails-plugin/build.gradle:
```diff
@@ -38,12 +37,15 @@ dependencies {
+
     testRuntimeOnly "com.h2database:h2"
     testRuntimeOnly "org.apache.tomcat:tomcat-jdbc"
+    testRuntimeOnly "org.hibernate:hibernate-ehcache:$hibernateVersion"
```
Add the following to document the right approach and avoid conflicts from the javax version?

```
, {
    exclude group:'org.hibernate', module:'hibernate-core'
}
```

jamesfredley commented:
@matrei The documentation updates on https://github.com/grails/grails-doc/pull/951 look great.  What do you think about the 3 small changes I commented on the code?

matrei committed:
build: explicitly exclude `hibernate-core` from `hibernate-ehcache`

Even if it does not cause issues in this project, we explicitly exclude
the `javax` variant of `hibernate-core` from `hibernate-ehcache` to
document best practices and prevent potential conflicts in other setups.

matrei commented:
@jamesfredley Well, that failed spectaculary 😄 . I'll check what's needed to make it run again.

jamesfredley commented:
@matrei It is the same error that I say initially.

The following does resolve it, although I still think it is suboptimal since it includes 19 javax.transaction.* classes.

```
// required for org.hibernate:hibernate-ehcache to work with org.hibernate:hibernate-core-jakarta
implementation "org.jboss.spec.javax.transaction:jboss-transaction-api_1.3_spec:$jbossTransactionApiVersion", {
    exclude group:"jakarta.enterprise", module:"jakarta.enterprise.cdi-api"
}
```

matrei commented:
> The following does resolve it, although I still think it is suboptimal since it includes 19 javax.transaction.* classes

@jamesfredley Yes, you are right. The question is where we put the responsibility to add this. I would like to say on the application as we have already concluded `hibernate-ehcache` is an application concern. But it can be added as a `runtimeOnly` dependency and I think the exclusion of `jakarta.enterprise.cdi-api` is not needed as it is marked as `provided`.

jamesfredley commented:
@matrei I agree it is an application concern and am onboard with reducing the scope when we can.  For this project we can add it  as needed, for the 3 `runtimeOnly/testRuntimeOnly` locations in the projects.

matrei committed:
build: add `jboss-transaction-api` for `hibernate-ehcache`

Excluding the `javax` variant from `hibernate-ehcache` requires explicitly
including `jboss-transaction-api`.

matrei committed:
docs: update GORM in Spring Boot section

- Revise the instructions for using GORM in a Spring Boot application.
- Simplify the `application.yml` file in the Spring Boot test project.

matrei commented:
> I agree it is an application concern and am onboard with reducing the scope when we can. For this project we can add it as needed, for the 3 runtimeOnly/testRuntimeOnly locations in the projects.

@jamesfredley Excellent! Changes made, please review.

jdaugherty commented:
I worry about this change.  The gorm documentation (under Getting Started) recommends ehcache be included.  Caching is a critical component of hibernate and we are definitely making this harder to include because of the extra gradle workarounds that are needed.  We need to be sure to update the gorm docs for these workarounds so people understand this is only until we have hibernate 6.

matrei commented:
> The gorm documentation (under Getting Started) recommends ehcache be included.

Yes, exactly! The documentation does not say that `hibernate-cache` is provided transitively. It says to include it explicitly.

> We need to be sure to update the gorm docs for these workarounds so people understand this is only until we have hibernate 6.

Yes, it looks very outdated and needs to be updated.

matrei committed:
build: fix wrong dependency configuration

jamesfredley commented:
I will also try to preemptively address the following, which will break with this change.

https://github.com/search?q=org%3Agrails+org.hibernate.cache.ehcache+language%3AYAML&type=code&l=YAML

jamesfredley commented:
https://github.com/grails/gorm-mongodb/pull/800

jamesfredley commented:
https://github.com/grails/grails-spring-security-core/pull/1055

jamesfredley commented:
https://github.com/grails/grails-core/pull/14032

jamesfredley commented:
https://github.com/grails/grails-views/pull/613
